File: src/search_engine.py
"""Simple motorcycle repair POI search using Nominatim + Overpass.

Usage example:
    python -m src.search_engine --country "Italy" --city "Milan" --radius 10000 --limit 25

This module is intended as a starting point. It queries public OSM services at runtime.
"""

import argparse
import csv
import json
import time
from typing import Dict, Any, List, Optional

import requests

USER_AGENT = "motorcycle-search-agent/1.0 (your-email@example.com)"
NOMINATIM_URL = "https://nominatim.openstreetmap.org/search"
OVERPASS_URL = "https://overpass-api.de/api/interpreter"


def load_cities(path: str) -> List[Dict[str, str]]:
    rows = []
    with open(path, newline='', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for r in reader:
            rows.append({"country": r['Country'].strip(), "city": r['City'].strip()})
    return rows


def geocode_city(city: str, country: str) -> Optional[Dict[str, Any]]:
    params = {"city": city, "country": country, "format": "json", "limit": 1}
    headers = {"User-Agent": USER_AGENT}
    resp = requests.get(NOMINATIM_URL, params=params, headers=headers, timeout=15)
    resp.raise_for_status()
    data = resp.json()
    if not data:
        return None
    return data[0]


def overpass_query(lat: float, lon: float, radius: int = 10000, limit: int = 50) -> List[Dict[str, Any]]:
    # Flexible Overpass query: look for nodes/ways/relations with tags or name keywords indicating motorcycle repair
    query = f"""
[out:json][timeout:25];
(
  node(around:{radius},{lat},{lon})[shop~"motorcycle|car_repair|garage|bicycle",i];
  way(around:{radius},{lat},{lon})[shop~"motorcycle|car_repair|garage|bicycle",i];
  relation(around:{radius},{lat},{lon})[shop~"motorcycle|car_repair|garage|bicycle",i];
  node(around:{radius},{lat},{lon})[name~"motorcycle|moto|motorbike|motorrad|mc",i];
  way(around:{radius},{lat},{lon})[name~"motorcycle|moto|motorbike|motorrad|mc",i];
  relation(around:{radius},{lat},{lon})[name~"motorcycle|moto|motorbike|motorrad|mc",i];
);
out center {limit};
""".strip()
    headers = {"User-Agent": USER_AGENT}
    resp = requests.post(OVERPASS_URL, data={'data': query}, headers=headers, timeout=60)
    resp.raise_for_status()
    data = resp.json()
    elements = data.get('elements', [])[:limit]
    results = []
    for e in elements:
        tags = e.get('tags', {})
        lat_e = e.get('lat') or (e.get('center') or {}).get('lat')
        lon_e = e.get('lon') or (e.get('center') or {}).get('lon')
        results.append({
            'id': e.get('id'),
            'type': e.get('type'),
            'lat': lat_e,
            'lon': lon_e,
            'tags': tags,
        })
    return results


def search_city(country: str, city: str, radius: int = 10000, limit: int = 50) -> Dict[str, Any]:
    geocoded = geocode_city(city, country)
    if not geocoded:
        return {"country": country, "city": city, "error": "geocode_failed"}
    lat = float(geocoded['lat'])
    lon = float(geocoded['lon'])
    time.sleep(1)  # be gentle to Nominatim
    pois = overpass_query(lat, lon, radius=radius, limit=limit)
    return {
        'country': country,
        'city': city,
        'lat': lat,
        'lon': lon,
        'count': len(pois),
        'pois': pois,
    }


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--country', required=False)
    parser.add_argument('--city', required=False)
    parser.add_argument('--radius', type=int, default=10000)
    parser.add_argument('--limit', type=int, default=50)
    parser.add_argument('--cities-file', default='data/cities.csv')
    args = parser.parse_args()

    if args.country and args.city:
        res = search_city(args.country, args.city, radius=args.radius, limit=args.limit)
        print(json.dumps(res, indent=2, ensure_ascii=False))
        return

    cities = load_cities(args.cities_file)
    results = []
    for c in cities:
        print(f"Searching {c['city']}, {c['country']}...")
        try:
            r = search_city(c['country'], c['city'], radius=args.radius, limit=args.limit)
        except Exception as e:
            r = {'country': c['country'], 'city': c['city'], 'error': str(e)}
        results.append(r)
        # gentle pause between city requests
        time.sleep(1)

    print(json.dumps(results, indent=2, ensure_ascii=False))


if __name__ == '__main__':
    main()

File: lib/search-engine.ts
import fs from 'fs';
import path from 'path';

const USER_AGENT = 'motorcycle-search-agent/1.0 (your-email@example.com)';
const NOMINATIM_URL = 'https://nominatim.openstreetmap.org/search';
const OVERPASS_URL = 'https://overpass-api.de/api/interpreter';

export interface City {
  country: string;
  city: string;
}

export interface GeocodedLocation {
  lat: string;
  lon: string;
  display_name?: string;
}

export interface POI {
  id: number;
  type: string;
  lat?: number;
  lon?: number;
  tags: Record<string, string>;
}

export interface SearchResult {
  country: string;
  city: string;
  lat?: number;
  lon?: number;
  count: number;
  pois: POI[];
  error?: string;
}

export function loadCities(filePath: string = 'data/cities.csv'): City[] {
  const fullPath = path.join(process.cwd(), filePath);
  const fileContent = fs.readFileSync(fullPath, 'utf-8');
  const lines = fileContent.split('\n').filter(line => line.trim());

  // Skip header
  const cities: City[] = [];
  for (let i = 1; i < lines.length; i++) {
    const [country, city] = lines[i].split(',').map(s => s.trim());
    if (country && city) {
      cities.push({ country, city });
    }
  }

  return cities;
}

export async function geocodeCity(city: string, country: string): Promise<GeocodedLocation | null> {
  const params = new URLSearchParams({
    city,
    country,
    format: 'json',
    limit: '1',
  });

  const response = await fetch(`${NOMINATIM_URL}?${params}`, {
    headers: { 'User-Agent': USER_AGENT },
  });

  if (!response.ok) {
    throw new Error(`Nominatim API error: ${response.status}`);
  }

  const data = await response.json();
  if (!data || data.length === 0) {
    return null;
  }

  return data[0] as GeocodedLocation;
}

export async function overpassQuery(
  lat: number,
  lon: number,
  radius: number = 10000,
  limit: number = 50
): Promise<POI[]> {
  const query = `
[out:json][timeout:25];
(
  node(around:${radius},${lat},${lon})[shop~"motorcycle|car_repair|garage|bicycle",i];
  way(around:${radius},${lat},${lon})[shop~"motorcycle|car_repair|garage|bicycle",i];
  relation(around:${radius},${lat},${lon})[shop~"motorcycle|car_repair|garage|bicycle",i];
  node(around:${radius},${lat},${lon})[name~"motorcycle|moto|motorbike|motorrad|mc",i];
  way(around:${radius},${lat},${lon})[name~"motorcycle|moto|motorbike|motorrad|mc",i];
  relation(around:${radius},${lat},${lon})[name~"motorcycle|moto|motorbike|motorrad|mc",i];
);
out center ${limit};
  `.trim();

  const formData = new URLSearchParams();
  formData.append('data', query);

  const response = await fetch(OVERPASS_URL, {
    method: 'POST',
    headers: {
      'User-Agent': USER_AGENT,
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: formData,
  });

  if (!response.ok) {
    throw new Error(`Overpass API error: ${response.status}`);
  }

  const data = await response.json();
  const elements = (data.elements || []).slice(0, limit);

  const results: POI[] = elements.map((e: any) => ({
    id: e.id,
    type: e.type,
    lat: e.lat || e.center?.lat,
    lon: e.lon || e.center?.lon,
    tags: e.tags || {},
  }));

  return results;
}

export async function searchCity(
  country: string,
  city: string,
  radius: number = 10000,
  limit: number = 50
): Promise<SearchResult> {
  try {
    const geocoded = await geocodeCity(city, country);

    if (!geocoded) {
      return {
        country,
        city,
        count: 0,
        pois: [],
        error: 'geocode_failed',
      };
    }

    const lat = parseFloat(geocoded.lat);
    const lon = parseFloat(geocoded.lon);

    // Be gentle to Nominatim - add a small delay
    await new Promise(resolve => setTimeout(resolve, 1000));

    const pois = await overpassQuery(lat, lon, radius, limit);

    return {
      country,
      city,
      lat,
      lon,
      count: pois.length,
      pois,
    };
  } catch (error) {
    return {
      country,
      city,
      count: 0,
      pois: [],
      error: error instanceof Error ? error.message : 'Unknown error',
    };
  }
}
